scheduler/src/cook/scheduler/constraints.clj

```diff
@@ -246,26 +251,51 @@
   (let [enforcing? (ratelimit/enforce? ratelimit/global-job-launch-rate-limiter)
         max-tasks (ratelimit/get-token-count! ratelimit/global-job-launch-rate-limiter ratelimit/global-job-launch-rate-limiter-key)]
     (when enforcing?
-      (reify com.netflix.fenzo.ConstraintEvaluator
+      (reify ConstraintEvaluator
         (getName [_] "launch_max_tasks")
         (evaluate [_ _ _ task-tracker-state]
           (let [num-assigned (-> task-tracker-state .getAllCurrentlyAssignedTasks .size)]
-            (com.netflix.fenzo.ConstraintEvaluator$Result.
+            (ConstraintEvaluator$Result.
               (< num-assigned max-tasks)
               (str "Hit the global rate limit"))))))))
 
+(defn build-max-tasks-per-host-constraint
+  "Returns a Fenzo constraint that ensures that we don't
+  match more tasks per host than is allowed (configured)"
+  []
+  (reify ConstraintEvaluator
+    (getName [_] "max_tasks_per_host")
+    (^ConstraintEvaluator$Result evaluate
+      [^ConstraintEvaluator _
+       ^TaskRequest _
+       ^VirtualMachineCurrentState target-vm
+       ^TaskTrackerState _]
+      (let [vm-resources (.getCurrAvailableResources target-vm)
+            vm-attributes (get-vm-lease-attr-map vm-resources)
+            max-per-host (get vm-attributes "COOK_MAX_TASKS_PER_HOST")]
+        (if max-per-host
+          (let [num-running (get vm-attributes "COOK_NUM_TASKS_ON_HOST")
+                num-assigned (-> target-vm .getTasksCurrentlyAssigned count)
+                num-total (+ num-running num-assigned)]
+            (ConstraintEvaluator$Result.
+              (< num-total max-per-host)
```

Comment:
Shouldn't this be `<=`?

Reply:
If `num-total == max-per-host`, then we want this constraint to return `false`, because it means the sum of running tasks and assigned tasks on this host has reached the max.

Reply:
OK, yes, I misunderstood what `num-assigned` means.